// ably/build/push.js
/*@license Copyright 2015-2022 Ably Real-time Ltd (ably.com)

Ably JavaScript Library v2.25.0
https://github.com/ably/ably-js

Released under the Apache Licence v2.0*/(function (g, f) {
    if ("object" == typeof exports && "object" == typeof module) {
      module.exports = f(require('ulid'));
    } else if ("function" == typeof define && define.amd) {
      define(['ulid'], f);
    } else if ("object" == typeof exports) {
      exports["AblyPushPlugin"] = f(require('ulid'));
    } else {
      g["AblyPushPlugin"] = f(g["ulid"]);
    }
  }(this, (__da) => {
var exports = {};
var module = { exports };
"use strict";
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/plugins/push/index.ts
var push_exports = {};
__export(push_exports, {
  ActivationStateMachine: () => ActivationStateMachine,
  CalledActivate: () => CalledActivate,
  CalledDeactivate: () => CalledDeactivate,
  PushChannel: () => pushchannel_default,
  default: () => push_default,
  getW3CPushDeviceDetails: () => getW3CPushDeviceDetails,
  localDeviceFactory: () => localDeviceFactory
});
module.exports = __toCommonJS(push_exports);

// src/plugins/push/pushchannel.ts
var PushChannel = class {
  constructor(channel) {
    this.channel = channel;
    this.client = channel.client;
  }
  async subscribeDevice() {
    const client = this.client;
    const device = await client.getDevice();
    const format = client.options.useBinaryProtocol ? client.Utils.Format.msgpack : client.Utils.Format.json, body = { deviceId: device.id, channel: this.channel.name }, headers = client.Defaults.defaultPostHeaders(client.options);
    if (client.options.headers)
      client.Utils.mixin(headers, client.options.headers);
    client.Utils.mixin(headers, this._getPushAuthHeaders(device));
    const requestBody = client.Utils.encodeBody(body, client._MsgPack, format);
    await client.rest.Resource.post(client, "/push/channelSubscriptions", requestBody, headers, {}, format, true);
  }
  async unsubscribeDevice() {
    const client = this.client;
    const device = await client.getDevice();
    const format = client.options.useBinaryProtocol ? client.Utils.Format.msgpack : client.Utils.Format.json, headers = client.Defaults.defaultPostHeaders(client.options);
    if (client.options.headers)
      client.Utils.mixin(headers, client.options.headers);
    client.Utils.mixin(headers, this._getPushAuthHeaders(device));
    await client.rest.Resource.delete(
      client,
      "/push/channelSubscriptions",
      headers,
      { deviceId: device.id, channel: this.channel.name },
      format,
      true
    );
  }
  async subscribeClient() {
    const client = this.client;
    const clientId = this.client.auth.clientId;
    if (!clientId) {
      throw new this.client.ErrorInfo({
        message: "Cannot subscribe from client without client ID",
        code: 5e4,
        statusCode: 500,
        remediation: "Set ClientOptions.clientId before calling pushChannel.subscribeClient(). On a realtime client, a clientId carried in the token also satisfies this once the connection has connected. On a REST client, only ClientOptions.clientId works."
      });
    }
    const format = client.options.useBinaryProtocol ? client.Utils.Format.msgpack : client.Utils.Format.json, body = { clientId, channel: this.channel.name }, headers = client.Defaults.defaultPostHeaders(client.options);
    if (client.options.headers)
      client.Utils.mixin(headers, client.options.headers);
    const requestBody = client.Utils.encodeBody(body, client._MsgPack, format);
    await client.rest.Resource.post(client, "/push/channelSubscriptions", requestBody, headers, {}, format, true);
  }
  async unsubscribeClient() {
    const client = this.client;
    const clientId = this.client.auth.clientId;
    if (!clientId) {
      throw new this.client.ErrorInfo({
        message: "Cannot unsubscribe from client without client ID",
        code: 5e4,
        statusCode: 500,
        remediation: "Set ClientOptions.clientId before calling pushChannel.unsubscribeClient(). On a realtime client, a clientId carried in the token also satisfies this once the connection has connected. On a REST client, only ClientOptions.clientId works."
      });
    }
    const format = client.options.useBinaryProtocol ? client.Utils.Format.msgpack : client.Utils.Format.json, headers = client.Defaults.defaultPostHeaders(client.options);
    if (client.options.headers)
      client.Utils.mixin(headers, client.options.headers);
    await client.rest.Resource.delete(
      client,
      "/push/channelSubscriptions",
      headers,
      { clientId, channel: this.channel.name },
      format,
      true
    );
  }
  async listSubscriptions(params) {
    this.client.Logger.logAction(
      this.client.logger,
      this.client.Logger.LOG_MICRO,
      "PushChannel.listSubscriptions()",
      "channel = " + this.channel.name
    );
    return this.client.push.admin.channelSubscriptions.list(__spreadProps(__spreadValues({}, params), {
      channel: this.channel.name,
      concatFilters: true
    }));
  }
  _getDeviceIdentityToken(device) {
    const deviceIdentityToken = device.deviceIdentityToken;
    if (deviceIdentityToken) {
      return deviceIdentityToken;
    } else {
      throw new this.client.ErrorInfo({
        message: "Cannot subscribe or unsubscribe this device without a deviceIdentityToken",
        code: 5e4,
        statusCode: 500,
        remediation: "Activate this device first by awaiting client.push.activate()."
      });
    }
  }
  _getPushAuthHeaders(device) {
    const deviceIdentityToken = this._getDeviceIdentityToken(device);
    return { "X-Ably-DeviceToken": deviceIdentityToken };
  }
};
var pushchannel_default = PushChannel;

// src/plugins/push/getW3CDeviceDetails.ts
function toBase64Url(arrayBuffer) {
  const buffer = new Uint8Array(arrayBuffer.slice(0, arrayBuffer.byteLength));
  return btoa(String.fromCharCode.apply(null, Array.from(buffer)));
}
function urlBase64ToBase64(base64String) {
  const padding = "=".repeat((4 - base64String.length % 4) % 4);
  const base64 = (base64String + padding).replace(/-/g, "+").replace(/_/g, "/");
  return base64;
}
function base64ToUint8Array(base64String) {
  const rawData = window.atob(base64String);
  const rawDataChars = [];
  for (let i = 0; i < rawData.length; i++) {
    rawDataChars.push(rawData[i].charCodeAt(0));
  }
  return Uint8Array.from(rawDataChars);
}
async function getW3CPushDeviceDetails(machine) {
  const GettingPushDeviceDetailsFailed2 = machine.GettingPushDeviceDetailsFailed;
  const GotPushDeviceDetails2 = machine.GotPushDeviceDetails;
  const { ErrorInfo: ErrorInfo2, Defaults } = machine.client;
  const permission = await Notification.requestPermission();
  if (permission !== "granted") {
    const err = permission === "denied" ? new ErrorInfo2({
      message: 'User denied permission to send notifications: browser notification permission is "denied"',
      code: 4e4,
      statusCode: 400,
      remediation: 'Tell the user to re-enable notifications for this site in their browser settings, then call push.activate() again to retry registration. A re-request will not prompt while the permission stays "denied".'
    }) : new ErrorInfo2({
      message: 'Notification permission prompt was dismissed without a choice: browser notification permission is "default"',
      code: 4e4,
      statusCode: 400,
      remediation: "Surface UI explaining the value of notifications, then call push.activate() again to retry registration. The browser will show the permission prompt again."
    });
    machine.handleEvent(new GettingPushDeviceDetailsFailed2(err));
    return;
  }
  const swUrl = machine.client.options.pushServiceWorkerUrl;
  if (!swUrl) {
    const err = new ErrorInfo2({
      message: "Missing ClientOptions.pushServiceWorkerUrl",
      code: 4e4,
      statusCode: 400,
      remediation: 'Set ClientOptions.pushServiceWorkerUrl to the path of your service worker (e.g. "/ably-push-sw.js") so the SDK can register it for web push.'
    });
    machine.handleEvent(new GettingPushDeviceDetailsFailed2(err));
    return;
  }
  try {
    const worker = await navigator.serviceWorker.register(swUrl);
    machine._pushManager = worker.pushManager;
    const headers = Defaults.defaultGetHeaders(machine.client.options, { format: "text" });
    const appServerKey = (await machine.client.rest.Resource.get(machine.client, "/push/publicVapidKey", headers, {}, null, true)).body;
    if (!worker.active) {
      await navigator.serviceWorker.ready;
    }
    const subscription = await worker.pushManager.subscribe({
      userVisibleOnly: true,
      applicationServerKey: base64ToUint8Array(urlBase64ToBase64(appServerKey))
    });
    const endpoint = subscription.endpoint;
    const [p256dh, auth] = [subscription.getKey("p256dh"), subscription.getKey("auth")];
    if (!p256dh || !auth) {
      throw new ErrorInfo2("Public key not found", 5e4, 500);
    }
    const device = await machine.client.getDevice();
    device.push.recipient = {
      transportType: "web",
      targetUrl: btoa(endpoint),
      publicVapidKey: appServerKey,
      encryptionKey: {
        p256dh: toBase64Url(p256dh),
        auth: toBase64Url(auth)
      }
    };
    device.persist();
    machine.handleEvent(new GotPushDeviceDetails2());
  } catch (err) {
    machine.handleEvent(
      new GettingPushDeviceDetailsFailed2(
        new ErrorInfo2("Failed to register service worker: " + err.message, 5e4, 500)
      )
    );
  }
}

// src/plugins/push/pushactivation.ts
var import_ulid = require("ulid");
var PUSH_ACTIVATION_NOT_AVAILABLE_HINT = "Run push.activate() in a browser environment with service worker support, or in React Native using the ably/react-native-push plugin. From a server, use client.push.admin instead. Call client.push.admin.publish(recipient, payload) to send to a device or clientId. Call client.push.admin.deviceRegistrations.save(device) to register a device record.";
var persistKeys = {
  deviceId: "ably.push.deviceId",
  deviceSecret: "ably.push.deviceSecret",
  deviceIdentityToken: "ably.push.deviceIdentityToken",
  pushRecipient: "ably.push.pushRecipient",
  activationState: "ably.push.activationState"
};
function loggedStorageWrites(client, op, writes) {
  const promises = writes.filter((result) => !!result && typeof result.then === "function");
  if (promises.length === 0) {
    return;
  }
  const allWritten = Promise.all(promises).then(() => {
  });
  allWritten.catch((err) => {
    client.Logger.logAction(
      client.logger,
      client.Logger.LOG_ERROR,
      "Push." + op,
      "failed to persist push state to storage: " + client.Utils.inspectError(err)
    );
  });
  return allWritten;
}
function localDeviceFactory(deviceDetails) {
  return class LocalDevice extends deviceDetails {
    constructor(rest) {
      super();
      this.push = {};
      this.rest = rest;
    }
    static load(rest) {
      const device = new LocalDevice(rest);
      device.loadPersisted();
      return device;
    }
    static async loadAsync(rest) {
      const device = new LocalDevice(rest);
      await device.loadPersistedAsync();
      return device;
    }
    async listSubscriptions() {
      if (!this.rest.pushConfig) {
        throw new this.rest.ErrorInfo({
          message: "Push activation is not available on this platform: it requires a browser environment with service worker support, or a React Native environment with the ably/react-native-push plugin",
          code: 4e4,
          statusCode: 400,
          remediation: PUSH_ACTIVATION_NOT_AVAILABLE_HINT
        });
      }
      if (!this.id) {
        throw new this.rest.ErrorInfo({
          message: "Device not activated",
          code: 4e4,
          statusCode: 400,
          remediation: "Call client.push.activate() and await its completion before listing subscriptions."
        });
      }
      if (!this.deviceIdentityToken) {
        throw new this.rest.ErrorInfo("Cannot list device subscriptions without deviceIdentityToken", 5e4, 500);
      }
      const client = this.rest, format = client.options.useBinaryProtocol ? client.Utils.Format.msgpack : client.Utils.Format.json, envelope = client.http.supportsLinkHeaders ? void 0 : format, headers = client.Defaults.defaultGetHeaders(client.options);
      client.Utils.mixin(headers, client.options.headers, { "X-Ably-DeviceToken": this.deviceIdentityToken });
      return new client.rest.PaginatedResource(client, "/push/channelSubscriptions", headers, envelope, async function(body, headers2, unpacked) {
        return client.rest.PushChannelSubscription.fromResponseBody(
          body,
          client._MsgPack,
          unpacked ? void 0 : format
        );
      }).get({ deviceId: this.id });
    }
    // keep in sync with loadPersistedAsync()
    loadPersisted() {
      var _a;
      const pushConfig = this.rest.pushConfig;
      if (!pushConfig) {
        throw new this.rest.ErrorInfo({
          message: "Push activation is not available on this platform: it requires a browser environment with service worker support, or a React Native environment with the ably/react-native-push plugin",
          code: 4e4,
          statusCode: 400,
          remediation: PUSH_ACTIVATION_NOT_AVAILABLE_HINT
        });
      }
      if (pushConfig.storageIsAsync) {
        throw new this.rest.ErrorInfo({
          message: "The local device cannot be loaded synchronously: push storage on this platform is asynchronous",
          code: 4e4,
          statusCode: 400,
          remediation: "Use await client.getDevice() instead of client.device(). device() reads storage synchronously and is deprecated."
        });
      }
      this.platform = pushConfig.platform;
      this.clientId = (_a = this.rest.auth.clientId) != null ? _a : void 0;
      this.formFactor = pushConfig.formFactor;
      this.id = pushConfig.storage.get(persistKeys.deviceId);
      if (this.id) {
        this.deviceSecret = pushConfig.storage.get(persistKeys.deviceSecret);
        this.deviceIdentityToken = JSON.parse(
          pushConfig.storage.get(persistKeys.deviceIdentityToken) || "null"
        );
        this.push.recipient = JSON.parse(pushConfig.storage.get(persistKeys.pushRecipient) || "null");
      } else {
        this.resetId();
      }
    }
    // keep in sync with loadPersisted(); awaiting a non-promise is a no-op, so this single
    // implementation serves both synchronous (web) and asynchronous (React Native) storage
    async loadPersistedAsync() {
      var _a, _b, _c;
      const pushConfig = this.rest.pushConfig;
      if (!pushConfig) {
        throw new this.rest.ErrorInfo({
          message: "Push activation is not available on this platform: it requires a browser environment with service worker support, or a React Native environment with the ably/react-native-push plugin",
          code: 4e4,
          statusCode: 400,
          remediation: PUSH_ACTIVATION_NOT_AVAILABLE_HINT
        });
      }
      this.platform = pushConfig.platform;
      this.clientId = (_a = this.rest.auth.clientId) != null ? _a : void 0;
      this.formFactor = pushConfig.formFactor;
      this.id = (_b = await pushConfig.storage.get(persistKeys.deviceId)) != null ? _b : void 0;
      if (this.id) {
        this.deviceSecret = (_c = await pushConfig.storage.get(persistKeys.deviceSecret)) != null ? _c : void 0;
        this.deviceIdentityToken = JSON.parse(
          await pushConfig.storage.get(persistKeys.deviceIdentityToken) || "null"
        );
        this.push.recipient = JSON.parse(await pushConfig.storage.get(persistKeys.pushRecipient) || "null");
      } else {
        await this.resetId();
      }
    }
    persist() {
      const pushConfig = this.rest.pushConfig;
      if (!pushConfig) {
        throw new this.rest.ErrorInfo({
          message: "Push activation is not available on this platform: it requires a browser environment with service worker support, or a React Native environment with the ably/react-native-push plugin",
          code: 4e4,
          statusCode: 400,
          remediation: PUSH_ACTIVATION_NOT_AVAILABLE_HINT
        });
      }
      const writes = [];
      if (this.id) {
        writes.push(pushConfig.storage.set(persistKeys.deviceId, this.id));
      }
      if (this.deviceSecret) {
        writes.push(pushConfig.storage.set(persistKeys.deviceSecret, this.deviceSecret));
      }
      if (this.deviceIdentityToken) {
        writes.push(pushConfig.storage.set(persistKeys.deviceIdentityToken, JSON.stringify(this.deviceIdentityToken)));
      }
      if (this.push.recipient) {
        writes.push(pushConfig.storage.set(persistKeys.pushRecipient, JSON.stringify(this.push.recipient)));
      }
      return loggedStorageWrites(this.rest, "LocalDevice.persist()", writes);
    }
    resetId() {
      this.id = (0, import_ulid.ulid)();
      this.deviceSecret = (0, import_ulid.ulid)();
      return this.persist();
    }
    getAuthDetails(rest, headers, params) {
      if (!this.deviceIdentityToken) {
        throw new this.rest.ErrorInfo("Unable to update device registration; no deviceIdentityToken", 5e4, 500);
      }
      if (this.rest.http.supportsAuthHeaders) {
        return {
          headers: rest.Utils.mixin(
            { authorization: "Bearer " + rest.Utils.toBase64(this.deviceIdentityToken) },
            headers
          ),
          params
        };
      } else {
        return { headers, params: rest.Utils.mixin({ access_token: this.deviceIdentityToken }, params) };
      }
    }
  };
}
var ActivationStateMachine = class {
  constructor(rest) {
    // serializes asynchronous activation-state writes; see persist()
    this._activationStateWrite = Promise.resolve();
    // exported for testing
    this.GettingPushDeviceDetailsFailed = GettingPushDeviceDetailsFailed;
    this.GotPushDeviceDetails = GotPushDeviceDetails;
    var _a;
    this.client = rest;
    this._pushConfig = rest.pushConfig;
    if (!((_a = this._pushConfig) == null ? void 0 : _a.storageIsAsync)) {
      this.current = new ActivationStates[this.pushConfig.storage.get(persistKeys.activationState) || "NotActivated"](null);
    }
    this.pendingEvents = [];
    this.handling = false;
  }
  /**
   * Resolves the persisted activation state when storage is asynchronous. Must be awaited
   * before the first handleEvent() call; a no-op with synchronous storage, where the
   * constructor has already resolved the state.
   */
  async ensureInitialized() {
    var _a;
    if (this.current) {
      return;
    }
    (_a = this._initPromise) != null ? _a : this._initPromise = (async () => {
      try {
        const persistedName = await this.pushConfig.storage.get(persistKeys.activationState);
        this.current = new ActivationStates[persistedName || "NotActivated"](null);
      } catch (err) {
        this._initPromise = void 0;
        throw err;
      }
    })();
    return this._initPromise;
  }
  get pushConfig() {
    if (!this._pushConfig) {
      throw new this.client.ErrorInfo({
        message: "This platform is not supported as a target of push notifications: push activation requires a browser environment with service worker support, or a React Native environment with the ably/react-native-push plugin",
        code: 4e4,
        statusCode: 400,
        remediation: PUSH_ACTIVATION_NOT_AVAILABLE_HINT
      });
    }
    return this._pushConfig;
  }
  persist() {
    if (!(this.current && isPersistentState(this.current))) {
      return;
    }
    const name = this.current.name;
    if (this.pushConfig.storageIsAsync) {
      this._activationStateWrite = this._activationStateWrite.then(
        () => loggedStorageWrites(this.client, "ActivationStateMachine.persist()", [
          this.pushConfig.storage.set(persistKeys.activationState, name)
        ])
      ).catch(() => {
      });
    } else {
      loggedStorageWrites(this.client, "ActivationStateMachine.persist()", [
        this.pushConfig.storage.set(persistKeys.activationState, name)
      ]);
    }
  }
  callUpdateRegistrationFailedCallback(reason) {
    if (this.updateFailedCallback) {
      this.updateFailedCallback(reason);
    } else {
      this.client.Logger.logAction(
        this.client.logger,
        this.client.Logger.LOG_ERROR,
        "UpdateRegistrationFailed",
        "Failed updating device push registration: " + this.client.Utils.inspectError(reason)
      );
    }
  }
  callCustomRegisterer(device, isNew) {
    var _a;
    (_a = this.registerCallback) == null ? void 0 : _a.call(this, device, (error, deviceRegistration) => {
      if (error) {
        if (isNew) {
          this.handleEvent(new GettingDeviceRegistrationFailed(error));
        } else {
          this.handleEvent(new SyncRegistrationFailed(error));
        }
        return;
      }
      if (!deviceRegistration) {
        const err = new this.client.ErrorInfo({
          message: "registerCallback did not return deviceRegistration",
          code: 4e4,
          statusCode: 400,
          remediation: "Your registerCallback must invoke its callback with (null, deviceRegistration)."
        });
        this.handleEvent(new GettingDeviceRegistrationFailed(err));
        return;
      }
      if (isNew) {
        this.handleEvent(new GotDeviceRegistration(deviceRegistration));
      } else {
        this.handleEvent(new RegistrationSynced());
      }
    });
  }
  callCustomDeregisterer(device) {
    var _a;
    (_a = this.deregisterCallback) == null ? void 0 : _a.call(this, device, (err) => {
      if (err) {
        this.handleEvent(new DeregistrationFailed(err));
        return;
      }
      this.handleEvent(new Deregistered());
    });
  }
  async updateRegistration() {
    const localDevice = await this.client.getDevice();
    if (this.registerCallback) {
      this.callCustomRegisterer(localDevice, false);
    } else {
      const client = this.client;
      const format = client.options.useBinaryProtocol ? this.client.Utils.Format.msgpack : this.client.Utils.Format.json, body = client.rest.DeviceDetails.fromLocalDevice(localDevice), headers = this.client.Defaults.defaultPostHeaders(this.client.options, { format }), params = {};
      if (client.options.headers) {
        this.client.Utils.mixin(headers, client.options.headers);
      }
      if (client.options.pushFullWait) {
        this.client.Utils.mixin(params, { fullWait: "true" });
      }
      const requestBody = this.client.Utils.encodeBody(body, client._MsgPack, format);
      const authDetails = localDevice.getAuthDetails(client, headers, params);
      try {
        const response = await this.client.rest.Resource.patch(
          client,
          "/push/deviceRegistrations",
          requestBody,
          authDetails.headers,
          authDetails.params,
          format,
          true
        );
        this.handleEvent(new GotDeviceRegistration(response.body));
      } catch (err) {
        this.handleEvent(new GettingDeviceRegistrationFailed(err));
      }
    }
  }
  async deregister() {
    const device = await this.client.getDevice();
    if (this.deregisterCallback) {
      this.callCustomDeregisterer(device);
    } else {
      const rest = this.client;
      const format = rest.options.useBinaryProtocol ? this.client.Utils.Format.msgpack : this.client.Utils.Format.json, headers = this.client.Defaults.defaultPostHeaders(rest.options), params = { deviceId: device.id };
      if (rest.options.headers)
        this.client.Utils.mixin(headers, rest.options.headers);
      const authDetails = device.getAuthDetails(this.client, headers, params);
      if (rest.options.pushFullWait)
        this.client.Utils.mixin(params, { fullWait: "true" });
      try {
        await this.client.rest.Resource.delete(
          rest,
          "/push/deviceRegistrations",
          authDetails.headers,
          authDetails.params,
          format,
          true
        );
        this.handleEvent(new Deregistered());
      } catch (err) {
        this.handleEvent(new DeregistrationFailed(err));
      }
    }
  }
  callActivatedCallback(err) {
    var _a;
    (_a = this.activatedCallback) == null ? void 0 : _a.call(this, err);
    delete this.activatedCallback;
  }
  callDeactivatedCallback(err) {
    var _a;
    (_a = this.deactivatedCallback) == null ? void 0 : _a.call(this, err);
    delete this.deactivatedCallback;
  }
  handleEvent(event) {
    if (this.handling) {
      this.client.Platform.Config.nextTick(() => {
        this.handleEvent(event);
      });
      return;
    }
    this.handling = true;
    this.client.Logger.logAction(
      this.client.logger,
      this.client.Logger.LOG_MAJOR,
      "Push.ActivationStateMachine.handleEvent()",
      "handling event " + event.name + " from " + this.current.name
    );
    let maybeNext = this.current.processEvent(this, event);
    if (!maybeNext) {
      this.client.Logger.logAction(
        this.client.logger,
        this.client.Logger.LOG_MAJOR,
        "Push.ActivationStateMachine.handleEvent()",
        "enqueing event: " + event.name
      );
      this.pendingEvents.push(event);
      this.handling = false;
      return;
    }
    this.client.Logger.logAction(
      this.client.logger,
      this.client.Logger.LOG_MAJOR,
      "Push.ActivationStateMachine.handleEvent()",
      "transition: " + this.current.name + " -(" + event.name + ")-> " + maybeNext.name
    );
    this.current = maybeNext;
    while (this.pendingEvents.length > 0) {
      const pending = this.pendingEvents[0];
      this.client.Logger.logAction(
        this.client.logger,
        this.client.Logger.LOG_MAJOR,
        "Push.ActivationStateMachine.handleEvent()",
        "attempting to consume pending event: " + pending.name
      );
      maybeNext = this.current.processEvent(this, pending);
      if (!maybeNext) {
        break;
      }
      this.pendingEvents.splice(0, 1);
      this.client.Logger.logAction(
        this.client.logger,
        this.client.Logger.LOG_MAJOR,
        "Push.ActivationStateMachine.handleEvent()",
        "transition: " + this.current.name + " -(" + pending.name + ")-> " + maybeNext.name
      );
      this.current = maybeNext;
    }
    this.persist();
    this.handling = false;
  }
};
var CalledActivate = class {
  constructor(machine, registerCallback) {
    this.name = "CalledActivate";
    if (registerCallback) {
      machine.registerCallback = registerCallback;
    }
    machine.persist();
  }
};
var CalledDeactivate = class {
  constructor(machine, deregisterCallback) {
    this.name = "CalledDeactivate";
    machine.deregisterCallback = deregisterCallback;
    machine.persist();
  }
};
var GotPushDeviceDetails = class {
  constructor() {
    this.name = "GotPushDeviceDetails";
  }
};
var GettingPushDeviceDetailsFailed = class {
  constructor(reason) {
    this.name = "GettingPushDeviceDetailsFailed";
    this.reason = reason;
  }
};
var GotDeviceRegistration = class {
  constructor(deviceRegistration) {
    this.name = "GotDeviceRegistration";
    this.tokenDetails = deviceRegistration.deviceIdentityToken;
  }
};
var GettingDeviceRegistrationFailed = class {
  constructor(reason) {
    this.name = "GettingDeviceRegistrationFailed";
    this.reason = reason;
  }
};
var RegistrationSynced = class {
  constructor() {
    this.name = "RegistrationSynced";
  }
};
var SyncRegistrationFailed = class {
  constructor(reason) {
    this.name = "SyncRegistrationFailed";
    this.reason = reason;
  }
};
var Deregistered = class {
  constructor() {
    this.name = "Deregistered";
  }
};
var DeregistrationFailed = class {
  constructor(reason) {
    this.name = "DeregistrationFailed";
    this.reason = reason;
  }
};
var ActivationState = class {
  constructor(name) {
    this.name = name;
  }
};
var NotActivated = class _NotActivated extends ActivationState {
  constructor() {
    super("NotActivated");
  }
  processEvent(machine, event) {
    var _a, _b;
    if (event instanceof CalledDeactivate) {
      machine.callDeactivatedCallback(null);
      return new _NotActivated();
    } else if (event instanceof CalledActivate) {
      const device = machine.client.device();
      if (device.deviceIdentityToken != null) {
        if (device.clientId && device.clientId !== machine.client.auth.clientId) {
          machine.handleEvent(
            new SyncRegistrationFailed(
              new machine.client.ErrorInfo("clientId not compatible with local device clientId", 61002, 400)
            )
          );
          return null;
        }
        machine.pendingEvents.push(event);
        return new WaitingForNewPushDeviceDetails();
      }
      if (device.push.recipient) {
        machine.pendingEvents.push(new GotPushDeviceDetails());
      } else if (machine.pushConfig.getPushDeviceDetails) {
        (_b = (_a = machine.pushConfig).getPushDeviceDetails) == null ? void 0 : _b.call(_a, machine);
      } else if (machine.pushConfig.platform === "browser" /* Browser */) {
        getW3CPushDeviceDetails(machine);
      } else {
        machine.handleEvent(
          new GettingPushDeviceDetailsFailed(
            new machine.client.ErrorInfo("No available implementation to get push device details", 5e4, 500)
          )
        );
      }
      return new WaitingForPushDeviceDetails();
    } else if (event instanceof GotPushDeviceDetails) {
      return new _NotActivated();
    }
    return null;
  }
};
var WaitingForPushDeviceDetails = class _WaitingForPushDeviceDetails extends ActivationState {
  constructor() {
    super("WaitingForPushDeviceDetails");
  }
  processEvent(machine, event) {
    if (event instanceof CalledActivate) {
      return new _WaitingForPushDeviceDetails();
    } else if (event instanceof CalledDeactivate) {
      machine.callDeactivatedCallback(null);
      return new NotActivated();
    } else if (event instanceof GotPushDeviceDetails) {
      const client = machine.client;
      const device = client.device();
      if (machine.registerCallback) {
        machine.callCustomRegisterer(device, true);
      } else {
        const format = client.options.useBinaryProtocol ? machine.client.Utils.Format.msgpack : machine.client.Utils.Format.json, body = client.rest.DeviceDetails.fromLocalDevice(device), headers = machine.client.Defaults.defaultPostHeaders(client.options, { format }), params = {};
        if (client.options.headers)
          machine.client.Utils.mixin(headers, client.options.headers);
        if (client.options.pushFullWait)
          machine.client.Utils.mixin(params, { fullWait: "true" });
        const requestBody = machine.client.Utils.encodeBody(body, client._MsgPack, format);
        machine.client.rest.Resource.post(client, "/push/deviceRegistrations", requestBody, headers, params, null, true).then((response) => {
          const deviceDetails = response.unpacked ? response.body : client.rest.DeviceDetails.fromResponseBody(response.body, client._MsgPack, format);
          machine.handleEvent(new GotDeviceRegistration(deviceDetails));
        }).catch((err) => {
          machine.handleEvent(new GettingDeviceRegistrationFailed(err));
        });
      }
      return new WaitingForDeviceRegistration();
    } else if (event instanceof GettingPushDeviceDetailsFailed) {
      machine.callActivatedCallback(event.reason);
      return new NotActivated();
    }
    return null;
  }
};
var WaitingForDeviceRegistration = class _WaitingForDeviceRegistration extends ActivationState {
  constructor() {
    super("WaitingForDeviceRegistration");
  }
  processEvent(machine, event) {
    if (event instanceof CalledActivate) {
      return new _WaitingForDeviceRegistration();
    } else if (event instanceof GotDeviceRegistration) {
      const device = machine.client.device();
      device.deviceIdentityToken = event.tokenDetails.token;
      device.persist();
      machine.callActivatedCallback(null);
      return new WaitingForNewPushDeviceDetails();
    } else if (event instanceof GettingDeviceRegistrationFailed) {
      machine.callActivatedCallback(event.reason);
      return new NotActivated();
    }
    return null;
  }
};
var WaitingForNewPushDeviceDetails = class _WaitingForNewPushDeviceDetails extends ActivationState {
  constructor() {
    super("WaitingForNewPushDeviceDetails");
  }
  processEvent(machine, event) {
    if (event instanceof CalledActivate) {
      machine.callActivatedCallback(null);
      return new _WaitingForNewPushDeviceDetails();
    } else if (event instanceof CalledDeactivate) {
      machine.deregister();
      return new WaitingForDeregistration(this);
    } else if (event instanceof GotPushDeviceDetails) {
      machine.updateRegistration();
      return new WaitingForRegistrationSync();
    }
    return null;
  }
};
var WaitingForRegistrationSync = class _WaitingForRegistrationSync extends ActivationState {
  constructor(triggeredByCalledActivate = false) {
    super("WaitingForRegistrationSync");
    this.triggeredByCalledActivate = triggeredByCalledActivate;
  }
  processEvent(machine, event) {
    if (event instanceof CalledActivate && !this.triggeredByCalledActivate) {
      machine.callActivatedCallback(null);
      return new _WaitingForRegistrationSync(true);
    } else if (event instanceof RegistrationSynced) {
      return new WaitingForNewPushDeviceDetails();
    } else if (event instanceof SyncRegistrationFailed) {
      machine.callUpdateRegistrationFailedCallback(event.reason);
      return new AfterRegistrationSyncFailed();
    }
    return null;
  }
};
var AfterRegistrationSyncFailed = class extends ActivationState {
  constructor() {
    super("AfterRegistrationSyncFailed");
  }
  processEvent(machine, event) {
    if (event instanceof CalledActivate || event instanceof GotPushDeviceDetails) {
      machine.updateRegistration();
      return new WaitingForRegistrationSync(event instanceof CalledActivate);
    } else if (event instanceof CalledDeactivate) {
      machine.deregister();
      return new WaitingForDeregistration(this);
    }
    return null;
  }
};
var WaitingForDeregistration = class _WaitingForDeregistration extends ActivationState {
  constructor(previousState) {
    super("WaitingForDeregistration");
    this.previousState = previousState;
  }
  processEvent(machine, event) {
    if (event instanceof CalledDeactivate) {
      return new _WaitingForDeregistration(this.previousState);
    } else if (event instanceof Deregistered) {
      const device = machine.client.device();
      delete device.deviceIdentityToken;
      delete device.push.recipient;
      loggedStorageWrites(machine.client, "WaitingForDeregistration.processEvent()", [
        machine.pushConfig.storage.remove(persistKeys.deviceIdentityToken),
        machine.pushConfig.storage.remove(persistKeys.pushRecipient)
      ]);
      device.resetId();
      machine.callDeactivatedCallback(null);
      return new NotActivated();
    } else if (event instanceof DeregistrationFailed) {
      machine.callDeactivatedCallback(event.reason);
      return this.previousState;
    }
    return null;
  }
};
var ActivationStates = {
  NotActivated,
  WaitingForPushDeviceDetails,
  WaitingForDeviceRegistration,
  WaitingForNewPushDeviceDetails,
  WaitingForRegistrationSync,
  AfterRegistrationSyncFailed,
  WaitingForDeregistration
};
function isPersistentState(state) {
  return state.name == "NotActivated" || state.name == "WaitingForNewPushDeviceDetails";
}

// src/plugins/push/index.ts
var push_default = {
  ActivationStateMachine,
  localDeviceFactory,
  CalledActivate,
  CalledDeactivate,
  PushChannel: pushchannel_default,
  getW3CPushDeviceDetails
};
if (typeof module.exports == "object" && typeof exports == "object") {
  var __cp = (to, from, except, desc) => {
    if ((from && typeof from === "object") || typeof from === "function") {
      for (let key of Object.getOwnPropertyNames(from)) {
        if (!Object.prototype.hasOwnProperty.call(to, key) && key !== except)
        Object.defineProperty(to, key, {
          get: () => from[key],
          enumerable: !(desc = Object.getOwnPropertyDescriptor(from, key)) || desc.enumerable,
        });
      }
    }
    return to;
  };
  module.exports = __cp(module.exports, exports);
}
return module.exports;
}))
